File: backend/src/engine/adaptation.js
'use strict';
/**
 * Adaptation engine — rule-based parameter tuning.
 * Mirrors Python engine/adaptation.py exactly.
 */
const crud   = require('../db/crud');
const logger = require('../logger');

function _clamp(value, min, max) {
  return Math.max(min, Math.min(max, value));
}

async function runAdaptation(window = 20) {
  const params = await crud.getCurrentParams();
  if (!params) return { error: 'No parameters found in DB' };

  const trades = await crud.getClosedTrades(window);
  if (trades.length < 5) {
    return { skipped: true, reason: `Only ${trades.length} closed trades — need at least 5` };
  }

  const wins   = trades.filter(t => t.result === 'WIN');
  const losses = trades.filter(t => t.result === 'LOSS');
  const winRate = (wins.length / trades.length) * 100;

  const grossProfit = wins.reduce((s, t)   => s + (t.pnl || 0), 0);
  const grossLoss   = Math.abs(losses.reduce((s, t) => s + (t.pnl || 0), 0)) || 0.0001;
  const profitFactor = grossProfit / grossLoss;

  const atrValues = trades.filter(t => t.atr_at_entry).map(t => t.atr_at_entry);
  const avgAtr    = atrValues.length ? atrValues.reduce((a, b) => a + b, 0) / atrValues.length : null;

  const actions   = [];
  const newParams = { ...params };

  // ── Rule 1: Win-rate based SL tuning ────────────────────────────────────
  if (winRate < 40) {
    const oldSl = newParams.stop_loss_pct;
    const newSl = _clamp(oldSl + 0.1, params.min_stop_loss_pct, params.max_stop_loss_pct);
    newParams.stop_loss_pct = Math.round(newSl * 1000) / 1000;
    if (newSl !== oldSl) {
      actions.push({ rule: 'win_rate_sl', detail: `Win rate ${winRate.toFixed(1)}% < 40% → SL ${oldSl}% → ${newSl}%` });
    }
  } else if (winRate > 65 && profitFactor > 1.5) {
    const oldSl = newParams.stop_loss_pct;
    const newSl = _clamp(oldSl - 0.05, params.min_stop_loss_pct, params.max_stop_loss_pct);
    newParams.stop_loss_pct = Math.round(newSl * 1000) / 1000;
    if (newSl !== oldSl) {
      actions.push({ rule: 'win_rate_sl_tight', detail: `Win rate ${winRate.toFixed(1)}% > 65% + PF ${profitFactor.toFixed(2)} → SL ${oldSl}% → ${newSl}%` });
    }
  }

  // ── Rule 2: Volatility-based ATR multiplier ──────────────────────────────
  if (avgAtr !== null) {
    const priceValues = trades.filter(t => t.entry_price).map(t => t.entry_price);
    const avgPrice    = priceValues.length ? priceValues.reduce((a, b) => a + b, 0) / priceValues.length : 1.0;
    const atrPct      = (avgAtr / avgPrice) * 100;
    const oldMult     = newParams.atr_multiplier;

    if (atrPct > 0.5) {
      const newMult = _clamp(oldMult + 0.1, params.min_atr_multiplier, params.max_atr_multiplier);
      newParams.atr_multiplier = Math.round(newMult * 100) / 100;
      if (newMult !== oldMult) {
        actions.push({ rule: 'volatility_high', detail: `ATR% ${atrPct.toFixed(3)}% > 0.5% → widen ATR mult ${oldMult} → ${newMult}` });
      }
    } else if (atrPct < 0.2) {
      const newMult = _clamp(oldMult - 0.1, params.min_atr_multiplier, params.max_atr_multiplier);
      newParams.atr_multiplier = Math.round(newMult * 100) / 100;
      if (newMult !== oldMult) {
        actions.push({ rule: 'volatility_low', detail: `ATR% ${atrPct.toFixed(3)}% < 0.2% → tighten ATR mult ${oldMult} → ${newMult}` });
      }
    }
  }

  // ── Rule 3: EMA length tuning ────────────────────────────────────────────
  if (winRate < 45 && trades.length >= window) {
    const oldFast = newParams.ema_fast;
    const oldSlow = newParams.ema_slow;
    const newFast = Math.round(_clamp(oldFast + 1, params.min_ema_fast, params.max_ema_fast));
    const newSlow = Math.round(_clamp(oldSlow + 2, params.min_ema_slow, params.max_ema_slow));
    newParams.ema_fast = newFast;
    newParams.ema_slow = newSlow;
    if (newFast !== oldFast || newSlow !== oldSlow) {
      actions.push({ rule: 'ema_tune', detail: `EMA signals failing (WR ${winRate.toFixed(1)}%) → EMA ${oldFast}/${oldSlow} → ${newFast}/${newSlow}` });
    }
  }

  // ── Rule 4: TP ratio improvement ─────────────────────────────────────────
  if (profitFactor < 1.0 && winRate >= 40) {
    const oldTp = newParams.take_profit_pct;
    const newTp = _clamp(oldTp + 0.1, params.min_take_profit_pct, params.max_take_profit_pct);
    newParams.take_profit_pct = Math.round(newTp * 1000) / 1000;
    if (newTp !== oldTp) {
      actions.push({ rule: 'tp_improve', detail: `Profit factor ${profitFactor.toFixed(2)} < 1.0 → TP ${oldTp}% → ${newTp}%` });
    }
  }

  // ── Save & log ────────────────────────────────────────────────────────────
  const reason = actions.length ? actions.map(a => a.detail).join('; ') : 'No changes needed';
  const pv     = await crud.saveParams(newParams, reason, 'AUTO');

  await crud.logAdaptation({
    trades_evaluated:   trades.length,
    win_rate:           Math.round(winRate * 100) / 100,
    profit_factor:      Math.round(profitFactor * 1000) / 1000,
    avg_atr:            avgAtr !== null ? Math.round(avgAtr * 1000000) / 1000000 : null,
    actions_taken:      JSON.stringify(actions),
    new_params_version: pv.version,
  });

  logger.info(`Adaptation complete: ${actions.length} changes → version ${pv.version}`);

  return {
    trades_evaluated:   trades.length,
    win_rate:           Math.round(winRate * 100) / 100,
    profit_factor:      Math.round(profitFactor * 1000) / 1000,
    actions,
    new_params_version: pv.version,
    new_params:         newParams,
  };
}

module.exports = { runAdaptation };

File: python-backend/app/services/adaptation.py
import json
from math import sqrt

from sqlalchemy.orm import Session

from .. import crud
from ..strategy.dtc import DEFAULT_PARAMS
from .runtime_settings import get_learning_settings


def _clamp(value: float, min_value: float, max_value: float) -> float:
    return max(min_value, min(max_value, value))


def _tiny_step(current: float, target_delta: float, max_change_pct: float) -> tuple[float, float]:
    max_abs_delta = abs(current) * (max_change_pct / 100.0)
    bounded_delta = _clamp(target_delta, -max_abs_delta, max_abs_delta)
    return current + bounded_delta, bounded_delta


def run_adaptation(db: Session, window: int = 20) -> dict:
    learning = get_learning_settings(db)
    params = crud.get_current_params(db) or DEFAULT_PARAMS.copy()
    all_closed = crud.get_closed_trades(db, 100000)
    last_adapt_count_raw = crud.get_setting(db, "last_adapt_closed_count")
    last_adapt_count = int(last_adapt_count_raw) if last_adapt_count_raw else 0
    if len(all_closed) - last_adapt_count < learning["adaptation_cooldown_trades"]:
        return {"skipped": True, "reason": "Adaptation cooldown active"}

    trades = all_closed[:window]
    if len(trades) < learning["adaptation_min_closed_trades"]:
        return {"skipped": True, "reason": f"Only {len(trades)} closed trades"}

    wins = [t for t in trades if t.result == "WIN"]
    losses = [t for t in trades if t.result == "LOSS"]
    win_rate = len(wins) / len(trades)
    gross_profit = sum(t.pnl or 0 for t in wins)
    gross_loss = abs(sum(t.pnl or 0 for t in losses)) or 1e-6
    profit_factor = gross_profit / gross_loss

    atr_values = [t.atr_at_entry for t in trades if t.atr_at_entry]
    avg_atr = sum(atr_values) / len(atr_values) if atr_values else None

    confidence = abs(win_rate - 0.5) + (max(0.0, profit_factor - 1.0) * 0.05)
    if confidence < learning["adaptation_confidence_threshold"]:
        return {"skipped": True, "reason": f"Confidence {confidence:.4f} below threshold"}

    lr = learning["adaptation_lr"]
    actions = []
    deltas = []
    new_params = params.copy()

    sl_signal = (0.5 - win_rate) * 0.2
    sl_target_delta = sl_signal * lr * 100
    new_sl, sl_delta = _tiny_step(float(params["stop_loss_pct"]), sl_target_delta, learning["adaptation_max_change_pct"])
    new_sl = _clamp(new_sl, params["min_stop_loss_pct"], params["max_stop_loss_pct"])
    if abs(new_sl - params["stop_loss_pct"]) > 0:
        new_params["stop_loss_pct"] = round(new_sl, 5)
        deltas.append(sl_delta)
        actions.append({"rule": "tiny_sl_update", "detail": f"SL {params['stop_loss_pct']} -> {new_params['stop_loss_pct']}"})

    tp_signal = (profit_factor - 1.0) * 0.02
    tp_target_delta = tp_signal * lr * 100
    for key in ["tp1_multiplier", "tp2_multiplier", "tp3_multiplier", "tp4_multiplier"]:
        current = float(params[key])
        next_value, delta = _tiny_step(current, tp_target_delta, learning["adaptation_max_change_pct"])
        next_value = _clamp(next_value, params["min_tp_multiplier"], params["max_tp_multiplier"])
        if abs(next_value - current) > 0:
            new_params[key] = round(next_value, 5)
            deltas.append(delta)

    if avg_atr:
        atr_signal = -0.01 if profit_factor > 1.0 else 0.01
        for key in ["ema_1", "ema_2", "ema_3", "ema_4", "ema_5", "ema_6"]:
            current = float(params[key])
            next_value, delta = _tiny_step(current, atr_signal * lr * current, learning["adaptation_max_change_pct"])
            if key == "ema_1":
                next_value = _clamp(next_value, params["min_ema_1"], params["max_ema_1"])
            if key == "ema_6":
                next_value = _clamp(next_value, params["min_ema_6"], params["max_ema_6"])
            next_value = round(next_value)
            if next_value != int(current):
                new_params[key] = int(next_value)
                deltas.append(delta)

    delta_magnitude = sqrt(sum(d * d for d in deltas)) if deltas else 0.0
    if not actions and not deltas:
        actions.append({"rule": "no_op", "detail": "No bounded change passed confidence and clip checks"})

    version_row = crud.save_params(
        db,
        new_params,
        reason="; ".join(a["detail"] for a in actions),
        trigger="AUTO",
        confidence_score=confidence,
        delta_magnitude=delta_magnitude,
    )
    crud.log_adaptation(
        db,
        {
            "trades_evaluated": len(trades),
            "win_rate": round(win_rate * 100, 2),
            "profit_factor": round(profit_factor, 3),
            "avg_atr": avg_atr,
            "actions_taken": json.dumps(actions),
            "new_params_version": version_row.version,
            "confidence_score": confidence,
            "delta_magnitude": delta_magnitude,
            "rollback_triggered": 0,
        },
    )
    crud.set_setting(db, "last_adapt_closed_count", str(len(all_closed)))
    return {
        "trades_evaluated": len(trades),
        "win_rate": round(win_rate * 100, 2),
        "profit_factor": round(profit_factor, 3),
        "confidence": round(confidence, 6),
        "actions": actions,
        "new_params_version": version_row.version,
        "new_params": new_params,
    }
